anchore_engine/clients/grype_wrapper.py

```diff
@@ -16,442 +16,467 @@
-    local_db_dir = _get_default_grype_db_dir_from_config()
-
-    # Move the archive
-    grype_db_archive_copied_file_location = _move_grype_db_archive(
-        grype_db_archive_local_file_location, local_db_dir
+    MISSING_GRYPE_DB_SESSION_ERROR_MESSAGE = (
+        "Cannot access missing grype_db_session. Reinitialize grype_db."
     )
 
-    # Unpack the archive
-    latest_grype_db_dir = _open_grype_db_archive(
-        grype_db_archive_copied_file_location, local_db_dir, version_name
-    )
-
-    # Remove the unpacked archive
-    _remove_grype_db_archive(grype_db_archive_copied_file_location)
-
-    # Return the full path to the grype db file
-    return latest_grype_db_dir
-
-
-def _init_latest_grype_db_engine(latest_grype_db_dir):
-    """
-    Create and return the sqlalchemy engine object
-    """
-    logger.info(
-        "Creating new db engine based on the grype_db at {}".format(latest_grype_db_dir)
-    )
-    latest_grype_db_file = os.path.join(latest_grype_db_dir, VULNERABILITY_FILE_NAME)
-    db_connect = "sqlite:///{}".format(latest_grype_db_file)
-    latest_grype_db_engine = sqlalchemy.create_engine(db_connect, echo=True)
-    return latest_grype_db_engine
-
-
-def _init_latest_grype_db_session(grype_db_engine):
-    """
-    Create and return the db session
-    """
-    logger.info(
-        "Creating new grype_db session from engine based on {}".format(
-            grype_db_engine.url
-        )
-    )
-    SessionMaker = sessionmaker(bind=grype_db_engine)
-    grype_db_session = SessionMaker()
-    return grype_db_session
-
-
-def _init_latest_grype_db(lastest_grype_db_archive: str, version_name: str):
-    """
-    Write the db string to file, create the engine, and create the session
-    Return the file and session
-    """
-    latest_grype_db_dir = _move_and_open_grype_db_archive(
-        lastest_grype_db_archive, version_name
-    )
-    latest_grype_db_engine = _init_latest_grype_db_engine(latest_grype_db_dir)
-    latest_grype_db_session = _init_latest_grype_db_session(latest_grype_db_engine)
-
-    # Return the dir, file, and engine
-    return latest_grype_db_dir, latest_grype_db_session
-
-
-def _remove_local_grype_db(grype_db_dir):
-    """
-    Remove old the local grype db file
-    """
-    if os.path.exists(grype_db_dir):
-        logger.info("Removing old grype_db at {}".format(grype_db_dir))
-        shutil.rmtree(grype_db_dir)
-    else:
-        logger.error(
-            "Failed to remove grype db at {} as it cannot be found.".format(
-                grype_db_dir
+    def __new__(cls):
+        if cls._grype_wrapper_instance is None:
+            cls._grype_wrapper_instance = super(GrypeWrapperSingleton, cls).__new__(cls)
+            cls._grype_db_dir_internal = None
+            cls._grype_db_session_internal = None
+            cls._grype_db_lock = rwlock.RWLockWrite()
+        return cls._grype_wrapper_instance
+
+    @classmethod
+    def get_instance(self):
```

Comment:
not super clear as to why this method is necessary, but it doesn't do any harm either.

Reply:
Mostly just for code clarity, to make it very obvious to the caller what they are getting (vs the constructor)